feat: snapshot. Enabling spawns via snapshot by default

jeffreyrainy opened the pull request.

jeffreyrainy committed:
feat: snapshot. Enabling spawns via snapshot by default

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkConfig.cs:
```diff
@@ -151,7 +151,7 @@ public class NetworkConfig
         /// Whether or not to enable Snapshot System for spawn and despawn commands. Working but experimental.
         /// </summary>
-        public bool UseSnapshotSpawn { get; } = false;
+        public bool UseSnapshotSpawn { get; } = true;
```
(surprised pikachu face)

jeffreyrainy committed:
Merge branch 'develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
feat: snapshot. Fixing the snapshot spawns mechanism to not rely on Singleton

jeffreyrainy committed:
feat: snapshot. Removing the default on for snapshot spawns

jeffreyrainy committed:
Merge branch 'develop' into feature/enable-snapshot-spawn-pre

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge branch 'feature/enable-snapshot-spawn-pre' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
test: adjusted HiddenVariableTests to separate spawning from value update. Helps pinpointing the issue with snapshot enabled

NoelStephensUnity committed:
fix

This fixes the issue with the UnityTransport updates breaking the ability to still use UNetTransport.

NoelStephensUnity committed:
refactor

Just minor refactoring of when we actually need the ENABLE_RELAY_SERVICE define.
Also some cleanup.

NoelStephensUnity committed:
refactor

removing the InputField pefab.
unpacked this inside of the ConnectionModeButtons prefab.
updated SampleScene to reflect this change

NoelStephensUnity committed:
Merge branch 'develop' into fix/unitytransport-connectionmode-buttons

NoelStephensUnity committed:
style

updating comments

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge remote-tracking branch 'origin/fix/unitytransport-connectionmode-buttons' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
chore: removal of EnableNetworkVariable in NetworkConfig. It's always True now

jeffreyrainy committed:
Merge branch 'develop' into chore/enable-network-variable-removal

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge remote-tracking branch 'origin/chore/enable-network-variable-removal' into feature/enable-snapshot-spawn

jeffreyrainy committed:
feat: snpashot enabled. Marking variables as dirty instead of serializing them with spawn command

jeffreyrainy committed:
fix: this bool write cannot be remoed yet

jeffreyrainy committed:
Merge remote-tracking branch 'origin/chore/enable-network-variable-removal' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
feat: snapshot spawn. Improving test, and re-enabling

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -391,6 +391,8 @@ internal void SendSpawnCallForObject(ulong clientId, NetworkObject networkObject
                     NetworkManager.NetworkMetrics.TrackObjectSpawnSent(clientId, networkObject.NetworkObjectId, networkObject.name, size);
                 }
+
+                networkObject.MarkVariablesDirty();
```
After spawning an object, we mark all its variable dirty. Trying to separate peas and carrots. Spawning is a thing. The value update is another.

0xFA11 replied:
wasn't inital netvars written to the payload next to spawn message?

jeffreyrainy replied:
Yes, sharp eyes. :-) That will be the final piece:

```
--- a/com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs
+++ b/com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs
@@ -453,8 +453,6 @@ namespace Unity.Netcode
                 var (isReparented, latestParent) = networkObject.GetNetworkParenting();
                 NetworkObject.WriteNetworkParenting(writer, isReparented, latestParent);
             }
-
-            networkObject.WriteNetworkVariableData(writer.GetStream(), clientId);
         }
```

It is not efficient to have such a large spawn message. The spawn itself is just "spawn this". That way, it can get efficiently stored, sent, ack'ed, re-sent, etc... by the snapshot system.
A different message, after the spawn, will carry the value update. It seems in current status, this separation works fine excpet for those Network{List,Set,Dictionary}. So, this PR is not ready. Still in progress.

jeffreyrainy replied:
To clarify, this change-to-be-done-later above, makes it symmetric. With it, both snapshot and non-snapshot have the spawns and the value updates separated.

0xFA11 replied:
so, are you saying that we're going to get rid of that initial netvar write block that writes into one single payload and have only the spawn message first, mark things dirty and leave the rest of the things up to netvar replication (whether snapshotted or not) to eventually update props on the clients later? if so, I do support that direction :) however, I wonder how many thing will break simply because of expecting initial netvar values to be there at the same time when we spawn the networkobject — let's break them by doing the right thing and fix them later when they come up!

0xFA11 reviewed (changes requested):
(changing my review status due to ongoing work)

jeffreyrainy committed:
feat: snapshot. Pre-requisite to enabling snapshot spawns. Separating spawn from value updates for NetworkList

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
Merge remote-tracking branch 'origin/develop' into feature/enable-snapshot-spawn-pre2

jeffreyrainy committed:
style: standards.py fix, removing useless using directive

jeffreyrainy committed:
feat: this branch should have had snapshot spawns enabled since the start.

jeffreyrainy committed:
Merge branch 'develop' into feature/enable-snapshot-spawn

jeffreyrainy committed:
feat: snapshot spawn. Allowing spawns in other stages

jeffreyrainy committed:
Merge branch 'feature/enable-snapshot-spawn' of github.com:Unity-Technologies/com.unity.multiplayer.mlapi into feature/enable-snapshot-spawn
